### ISSUE TITLE
Re-authentication fails with OAUTHBEARER when roleArn is used in the default credential

### ISSUE THREAD

**Issue by BDeus** [aws/aws-msk-iam-sasl-signer-js#19]
### Checkboxes for prior research

- [X] I've gone through [Developer Guide](https://docs.aws.amazon.com/msk/latest/developerguide/what-is-msk.html) and [API reference](https://docs.aws.amazon.com/msk/1.0/apireference/what-is-msk.html)
- [X] I've searched for [previous similar issues](https://github.com/aws/aws-msk-iam-sasl-signer-js/issues) and didn't find any solution.

### Describe the bug

Hi,
i think there are the same issue as the Java one before[ fixing it](https://github.com/aws/aws-msk-iam-auth/issues/143) in 2.0.2 
`SASL OAUTHBEARER authentication failed: Cannot change principals during re-authentication from IAM.arn:aws:sts::xxxx:assumed-role/xxxx/276eccf3-xxxx-role: IAM.arn:aws:sts::xxxx:assumed-role/xxxx/b489ee67-xxxx-role`

Could we apply the same logic (refreshing of credentials before Oauth token generation) ?

### aws-msk-iam-sasl-signer-js library version used

1.0.0

### Which Node.js version is this issue in?

20.10.0

### Operating System and version

linux

### Reproduction Steps

Using default credential provider
``` 
async oauthBearerTokenProvider(region: string) {
    // Uses AWS Default Credentials Provider Chain to fetch credentials
    const authTokenResponse = await generateAuthToken({ region });
    return {
        value: authTokenResponse.token,
    };
}
```

### Observed Behavior

With kafkajs and aws-msk-iam-sasl-signer-js 
`SASL OAUTHBEARER authentication failed: Cannot change principals during re-authentication from IAM.arn:aws:sts::xxxx:assumed-role/xxxx/276eccf3-xxxx-role: IAM.arn:aws:sts::xxxx:assumed-role/xxxx/b489ee67-xxxx-role`

### Expected Behavior

Avoid KafkaJSNonRetriableError when refreshing credentials

### Possible Solution

Refreshing of credentials before Oauth token generation

**Comment by agarwal1510** [comment]
Thank you for raising this. We will take a look and get back.

**Comment by jeevchiran** [comment]
Any further updates on this issue? or possible workaround ?

**Comment by sankalpbhatia** [comment]
Do we know which credentials provider is being used here via the default credentials chain?  Are these the same roles but with different "assumed role arns", or a different role? 

You can try using the ``awsDebugCreds`` [feature](https://github.com/aws/aws-msk-iam-sasl-signer-js?tab=readme-ov-file#finding-out-which-identity-is-being-used) we provide in this library to figure this out. The workaround should be to ensure in some way the arn of the caller identity is same in both cases. 

Nevertheless, for us to investigate further, it will be helpful if you can share what credentials provider is being used here to generate the tokens.

**Comment by BDeus** [comment]
Hi, for my case,
it's the same role that is been used (same Role ARN) but the RoleSessionName change.
example:
-  `IAM.arn:aws:sts::#ACCOUNT#:assumed-role/my-custom-role/339bc935-my-custom-role`
- `IAM.arn:aws:sts::#ACCOUNT#:assumed-role/my-custom-role/c0a362f1-my-customv-role`

**Comment by sankalpbhatia** [comment]
Would it be possible to share debug logs of the kafka client? One way to workaround this is to ensure the "role session name" you mention shouldn't change across tokens. We are interested in knowing which credentials provider in the credentials provider chain is being used for fetching the credentials. Once we know that, we can investigate further.

**Comment by sujayvenaik** [comment]
@sankalpbhatia is there any update on the fix here? we are also facing the same issue.

**Comment by sujayvenaik** [comment]
@BDeus, have you solved this issue at your end or found a temporary workaround?

**Comment by jeevchiran** [comment]
@sujayvenaik We tried one workaround, setting the env variable AWS_ROLE_SESSION_NAME with any random value, which solves the problem

**Comment by sujayvenaik** [comment]
@jeevchiran Hey! I think this is working for us on Kubernetes based compute. But is still failing  for us when we use this in Elastic Beanstalk (ec2 instances). 

What kind of compute were you using?

**Comment by sankalpbhatia** [comment]
@sujayvenaik apologies for not responding earlier. Would it be possible to share client level debug logs? If you can also share the steps required to repro this on our end, that would be helpful too. 

Thanks

**Comment by BDeus** [comment]
> @BDeus, have you solved this issue at your end or found a temporary workaround?

Hi,
for now we restart the consumer when exception is thrown.
It's obvious that is a K8S IAM issue for our part (no error is thrown when run directly on VM EC2)
We still use the old `kube2iam` project to deal with iam for K8S container. We will migrate to "IAM roles for service accounts" and see if this fix the issue.
I see that some other clients as resolve this issue with double authentication on default credential
https://github.com/aws/aws-msk-iam-auth/issues/143

**Comment by archepro84** [comment]
@sabper 

I have deployed the Nest.js framework on EKS and used IRSA to obtain permissions for MSK. However, the same issue still occurs.
To temporarily resolve this, I added `AWS_ROLE_SESSION_NAME` as an environment variable, but I believe this is not a fundamental solution.

**Comment by sabper** [comment]
@archepro84 
I am also in the same situation where nest.js was deployed by eks, and the situation was temporarily resolved by adding AWS_ROLE_SESSION_NAME.
I also think that we need a solution to the root cause.

**Comment by BDeus** [comment]
> @BDeus Was that issue fixed after migrating to IRSA? 👀 This issue is occurring even though we are already using IRSA.

no, it not solve the issue. we still need to override the environment variable

**Comment by mahwy** [comment]
You can override `roleSessionName` this way:

```js
import { fromNodeProviderChain } from '@aws-sdk/credential-providers';
import { generateAuthTokenFromCredentialsProvider } from 'aws-msk-iam-sasl-signer-js';

async function oauthBearerTokenProvider(region, sessionName) {
  const authTokenResponse = await generateAuthTokenFromCredentialsProvider({
    region,
    awsCredentialsProvider: fromNodeProviderChain({
      maxRetries: 3,
     // Override RoleSessionName
      roleSessionName: sessionName,
      clientConfig: {
        region,
      },
    }),
  });
  return {
    value: authTokenResponse.token,
  };
}
```
